Deployment of MicoServiceInterface fails in `mico-core`

Creation of Kubernetes Service fails.

Log:
```
2019-02-01 09:12:47.422 DEBUG 1 --- [pool-1-thread-1] i.g.u.m.core.REST.DeploymentController   : Creating Kubernetes deployment for MICO service 'hello' in version 'v1.0.0' 2019-02-01 09:12:47.878 ERROR 1 --- [pool-1-thread-1] i.g.u.m.core.REST.DeploymentController   : java.lang.NullPointerException

java.util.concurrent.CompletionException: java.lang.NullPointerException
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273) ~[na:1.8.0_181]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280) ~[na:1.8.0_181]
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:659) ~[na:1.8.0_181]
        at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:632) ~[na:1.8.0_181]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474) ~[na:1.8.0_181]
        at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1595) ~[na:1.8.0_181]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_181]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_181]
        at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_181]
Caused by: java.lang.NullPointerException: null
        at io.github.ust.mico.core.mapping.MicoKubernetesClient.createContainerPorts(MicoKubernetesClient.java:126) ~[classes!/:1.0-SNAPSHOT]
        at io.github.ust.mico.core.mapping.MicoKubernetesClient.createMicoService(MicoKubernetesClient.java:73) ~[classes!/:1.0-SNAPSHOT]
        at io.github.ust.mico.core.REST.DeploymentController.createKubernetesResources(DeploymentController.java:122) ~[classes!/:1.0-SNAPSHOT]
        at io.github.ust.mico.core.REST.DeploymentController.lambda$null$1(DeploymentController.java:84) ~[classes!/:1.0-SNAPSHOT]
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:656) ~[na:1.8.0_181]
        ... 6 common frames omitted
```

--- a/mico-core/src/main/java/io/github/ust/mico/core/REST/DeploymentController.java
+++ b/mico-core/src/main/java/io/github/ust/mico/core/REST/DeploymentController.java
@@ -24,7 +24,6 @@ import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RestController;
 import org.springframework.web.server.ResponseStatusException;
 
-import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
@@ -71,12 +70,17 @@ public class DeploymentController {
         }
         MicoApplication micoApplication = micoApplicationOptional.get();
 
-        List<MicoService> micoServices = micoApplication.getServices();
-        micoServices.forEach(micoService -> {
+        log.info("MicoApplication '{}' in version '{}' includes {} MicoService(s).",
+            shortName, version, micoApplication.getServices().size());
+
+        for (MicoService micoService : micoApplication.getServices()) {
+
             // TODO Check if build is already running -> no build required
             // TODO Check if image for the requested version is already in docker registry -> no build required
+
+            log.info("Start build of MicoService '{}' in version '{}'.", micoService.getShortName(), micoService.getVersion());
             backgroundTaskFactory.runAsync(() -> buildImageAndWait(micoService), dockerImageUri -> {
-                log.info("Build of image for service '{}' in version '{}' finished: {}",
+                log.info("Build of MicoService '{}' in version '{}' finished with image '{}'.",
                     micoService.getShortName(), micoService.getVersion(), dockerImageUri);
 
                 micoService.setDockerImageUri(dockerImageUri);
@@ -89,8 +93,7 @@ public class DeploymentController {
                     exceptionHandler(kre);
                 }
             }, this::exceptionHandler);
-        });
-
+        }
         return ResponseEntity.ok().build();
     }
 
@@ -105,7 +108,7 @@ public class DeploymentController {
                 return imageBuilder.createImageName(micoService.getShortName(), micoService.getVersion());
             } else {
                 booleanCompletableFuture.cancel(true);
-                throw new ImageBuildException("Build for service " + micoService.getShortName() + " failed");
+                throw new ImageBuildException("Build for service " + micoService.getShortName() + " in version " + micoService.getVersion() + " failed");
             }
         } catch (NotInitializedException | InterruptedException | ExecutionException | ImageBuildException | TimeoutException e) {
             log.error(e.getMessage(), e);
@@ -135,18 +138,19 @@ public class DeploymentController {
             log.warn("MICO application '{}' in version '{}' doesn't have any service deployment information stored.",
                 micoApplication.getShortName(), micoApplication.getShortName());
         }
-        log.debug("Creating Kubernetes deployment for MICO service '{}' in version '{}'",
+        log.info("Creating Kubernetes deployment for MicoService '{}' in version '{}'",
             micoService.getShortName(), micoService.getVersion());
+        log.debug("Details of MicoService: {}", micoService.toString());
         micoKubernetesClient.createMicoService(micoService, micoServiceDeploymentInfo);
 
-        log.debug("Creating {} Kubernetes service(s) for MICO service '{}' in version '{}'",
+        log.debug("Creating {} Kubernetes service(s) for MicoService '{}' in version '{}'",
             micoService.getServiceInterfaces().size(), micoService.getShortName(), micoService.getVersion());
         // Kubernetes Service(s)
         for (MicoServiceInterface serviceInterface : micoService.getServiceInterfaces()) {
             micoKubernetesClient.createMicoServiceInterface(serviceInterface, micoService);
         }
 
-        log.info("Created Kubernetes resources for MICO service '{}' in version '{}'",
+        log.info("Created Kubernetes resources for MicoService '{}' in version '{}'",
             micoService.getShortName(), micoService.getVersion());
     }
 

--- a/mico-core/src/main/java/io/github/ust/mico/core/imagebuilder/ImageBuilder.java
+++ b/mico-core/src/main/java/io/github/ust/mico/core/imagebuilder/ImageBuilder.java
@@ -193,7 +193,7 @@ public class ImageBuilder {
         build.setMetadata(metadata);
 
         Build createdBuild = buildClient.createOrReplace(build);
-        log.info("Build created");
+        log.info("Build created with name '{}'", buildName);
         log.debug("Created build: {} ", createdBuild);
 
         return createdBuild;

--- a/mico-core/src/test/java/io/github/ust/mico/core/DeploymentControllerTests.java
+++ b/mico-core/src/test/java/io/github/ust/mico/core/DeploymentControllerTests.java
@@ -1,28 +1,14 @@
 package io.github.ust.mico.core;
 
-import static io.github.ust.mico.core.TestConstants.DESCRIPTION;
-import static io.github.ust.mico.core.TestConstants.DOCKERFILE;
-import static io.github.ust.mico.core.TestConstants.DOCKER_IMAGE_URI;
-import static io.github.ust.mico.core.TestConstants.GIT_TEST_REPO_URL;
-import static io.github.ust.mico.core.TestConstants.ID;
-import static io.github.ust.mico.core.TestConstants.RELEASE;
-import static io.github.ust.mico.core.TestConstants.SHORT_NAME;
-import static io.github.ust.mico.core.TestConstants.VERSION;
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertNotNull;
-import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.BDDMockito.given;
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.verify;
-import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
-import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
-import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
-
-import java.util.Optional;
-import java.util.concurrent.CompletableFuture;
-import java.util.function.Consumer;
-import java.util.function.Function;
-
+import io.fabric8.kubernetes.api.model.apps.Deployment;
+import io.github.ust.mico.core.REST.DeploymentController;
+import io.github.ust.mico.core.concurrency.MicoCoreBackgroundTaskFactory;
+import io.github.ust.mico.core.imagebuilder.ImageBuilder;
+import io.github.ust.mico.core.mapping.MicoKubernetesClient;
+import io.github.ust.mico.core.model.*;
+import io.github.ust.mico.core.persistence.MicoApplicationRepository;
+import io.github.ust.mico.core.persistence.MicoServiceRepository;
+import io.github.ust.mico.core.util.CollectionUtils;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -38,22 +24,27 @@ import org.springframework.boot.test.mock.mockito.MockBean;
 import org.springframework.test.context.junit4.SpringRunner;
 import org.springframework.test.web.servlet.MockMvc;
 
-import io.fabric8.kubernetes.api.model.apps.Deployment;
-import io.github.ust.mico.core.REST.DeploymentController;
-import io.github.ust.mico.core.concurrency.MicoCoreBackgroundTaskFactory;
-import io.github.ust.mico.core.imagebuilder.ImageBuilder;
-import io.github.ust.mico.core.mapping.MicoKubernetesClient;
-import io.github.ust.mico.core.model.MicoApplication;
-import io.github.ust.mico.core.model.MicoService;
-import io.github.ust.mico.core.model.MicoServiceDeploymentInfo;
-import io.github.ust.mico.core.persistence.MicoApplicationRepository;
-import io.github.ust.mico.core.persistence.MicoServiceRepository;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static io.github.ust.mico.core.TestConstants.*;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.BDDMockito.given;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
 
 @RunWith(SpringRunner.class)
 @WebMvcTest(DeploymentController.class)
 @OverrideAutoConfiguration(enabled = true) //Needed to override our neo4j config
 @EnableAutoConfiguration
-@EnableConfigurationProperties(value = { CorsConfig.class })
+@EnableConfigurationProperties(value = {CorsConfig.class})
 public class DeploymentControllerTests {
 
     private static final String BASE_PATH = "/applications";
@@ -64,6 +55,8 @@ public class DeploymentControllerTests {
     ArgumentCaptor<Function<Throwable, Void>> onErrorArgumentCaptor;
     @Captor
     ArgumentCaptor<MicoService> micoServiceArgumentCaptor;
+    @Captor
+    ArgumentCaptor<MicoServiceInterface> micoServiceInterfaceArgumentCaptor;
     @Captor
     ArgumentCaptor<MicoServiceDeploymentInfo> deploymentInfoArgumentCaptor;
 
@@ -92,7 +85,7 @@ public class DeploymentControllerTests {
     }
 
     @Test
-    public void deployApplicationWithOneService() throws Exception {
+    public void deployApplicationWithOneServiceAndOneServiceInterface() throws Exception {
         MicoService service = getTestService();
         MicoApplication application = getTestApplication(service);
 
@@ -131,6 +124,20 @@ public class DeploymentControllerTests {
         int actualReplicas = deploymentInfo.getReplicas();
         int expectedReplicas = application.getDeploymentInfo().getServiceDeploymentInfos().get(service.getId()).getReplicas();
         assertEquals("Replicas does not match the definition in the deployment info", expectedReplicas, actualReplicas);
+
+        verify(micoKubernetesClient, times(1)).createMicoServiceInterface(
+            micoServiceInterfaceArgumentCaptor.capture(),
+            micoServiceArgumentCaptor.capture());
+
+        MicoServiceInterface micoServiceInterfaceToCreate = micoServiceInterfaceArgumentCaptor.getValue();
+        assertNotNull(micoServiceInterfaceToCreate);
+        assertEquals("MicoServiceInterface that will be created as Kubernetes resources does not match",
+            service.getServiceInterfaces().get(0), micoServiceInterfaceToCreate);
+
+        MicoService micoServiceThatIsUsedForInterfaceCreation = micoServiceArgumentCaptor.getValue();
+        assertNotNull(micoServiceThatIsUsedForInterfaceCreation);
+        assertEquals("MicoService that will be used to create a MicoServiceInterface does not match",
+            service, micoServiceThatIsUsedForInterfaceCreation);
     }
 
     private MicoApplication getTestApplication(MicoService service) {
@@ -139,7 +146,7 @@ public class DeploymentControllerTests {
             .setVersion(VERSION)
             .setDescription(DESCRIPTION);
         application.getDeploymentInfo().getServiceDeploymentInfos().put(ID, new MicoServiceDeploymentInfo()
-                .setReplicas(1));
+            .setReplicas(1));
         application.getServices().add(service);
         return application;
     }
@@ -150,6 +157,15 @@ public class DeploymentControllerTests {
             .setShortName(SHORT_NAME)
             .setVersion(RELEASE)
             .setGitCloneUrl(GIT_TEST_REPO_URL)
-            .setDockerfilePath(DOCKERFILE);
+            .setDockerfilePath(DOCKERFILE)
+            .setServiceInterfaces(CollectionUtils.listOf(
+                new MicoServiceInterface()
+                    .setServiceInterfaceName(SERVICE_INTERFACE_NAME)
+                    .setPorts(CollectionUtils.listOf(
+                        new MicoServicePort()
+                            .setNumber(80)
+                            .setTargetPort(80)
+                    ))
+            ));
     }
 }

--- a/mico-core/src/test/java/io/github/ust/mico/core/TestConstants.java
+++ b/mico-core/src/test/java/io/github/ust/mico/core/TestConstants.java
@@ -43,6 +43,7 @@ public class TestConstants {
     public static final String DESCRIPTION_2_MATCHER = JsonPathBuilder.buildSingleMatcher(DESCRIPTION_ATTRIBUTE, DESCRIPTION_2);
     public static final String DESCRIPTION_3_MATCHER = JsonPathBuilder.buildSingleMatcher(DESCRIPTION_ATTRIBUTE, DESCRIPTION_3);
 
+    public static final String SERVICE_INTERFACE_NAME = "service-interface-name";
 
     /**
      * Git repository that is used for testing.

